plugins/uefi-capsule/fu-uefi-bootmgr.c

```diff
@@ -249,6 +249,151 @@ fu_uefi_setup_bootnext_with_loadopt(FuEfiLoadOption *loadopt, GError **error)
+	if (sbatlevel_section == NULL) {
+		g_debug("no sbatlevel section was found");
+		/* if there is no .sbatlevel section, then it will not update, it should be safe */
+		if (g_error_matches(error_local, FWUPD_ERROR, FWUPD_ERROR_NOT_FOUND))
+			return TRUE;
+		g_propagate_error(error, g_steal_pointer(&error_local));
+	}
+
+	previous_sbatlevel = fu_firmware_get_image_by_id(sbatlevel_section, "previous", error);
+	if (previous_sbatlevel == NULL)
+		return FALSE;
+
+	shim_entries = fu_firmware_get_images(previous_sbatlevel);
+
+	current_sbatlevel_bytes =
+	    fu_efivar_get_data_bytes(FU_EFIVAR_GUID_SHIM, "SbatLevelRT", NULL, error);
+	if (current_sbatlevel_bytes == NULL) {
+		/* not safe if variable is not set but new shim would set it */
+		return FALSE;
+	}
+
+	current_sbatlevel = fu_csv_firmware_new();
+	fu_csv_firmware_add_column_id(FU_CSV_FIRMWARE(current_sbatlevel), "$id");
+	fu_csv_firmware_add_column_id(FU_CSV_FIRMWARE(current_sbatlevel), "component_generation");
+	fu_csv_firmware_add_column_id(FU_CSV_FIRMWARE(current_sbatlevel), "date_stamp");
+
+	if (!fu_firmware_parse(current_sbatlevel,
+			       current_sbatlevel_bytes,
+			       FWUPD_INSTALL_FLAG_NONE,
+			       error))
+		return FALSE;
+
+	/*
+	 * For every new shim entry, we need a matching entry in the
+	 * current sbatlevel. That is the entry of the shim is not
+	 * newer than current sbatlevel.
+	 *
+	 * The opposite way might work (for example shim's latest
+	 * sbatlevel matches) or not (shim is too old), but it will
+	 * not brick the current OS.
+	 */
+	for (idx = 0; idx < shim_entries->len; idx++) {
+		const gchar *entry_id = NULL;
+		guint64 shim_generation = 0;
+		guint64 current_generation = 0;
+		FuCsvEntry *current_entry = NULL;
+		FuCsvEntry *shim_entry = g_ptr_array_index(shim_entries, idx);
+
+		entry_id = fu_firmware_get_id(FU_FIRMWARE(shim_entry));
+
+		current_entry =
+		    FU_CSV_ENTRY(fu_firmware_get_image_by_id(FU_FIRMWARE(current_sbatlevel),
+							     entry_id,
+							     &error_local));
+		if (current_entry == NULL) {
+			if (g_error_matches(error_local, FWUPD_ERROR, FWUPD_ERROR_NOT_FOUND)) {
+				g_set_error(
+				    error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INVALID_FILE,
+				    "shim sbatlevel for %s has a bricking update for entry %s "
+				    "(missing entry in current UEFI variable)",
+				    source_shim,
+				    entry_id);
+			} else {
+				g_prefix_error(&error_local,
+					       "while looking for entry in current sbatlevel: ");
+				g_propagate_error(error, g_steal_pointer(&error_local));
+			}
+
+			return FALSE;
+		}
+
+		if (!fu_uefi_bootmgr_get_generation(shim_entry, &shim_generation, error)) {
+			g_prefix_error(error,
+				       "sbatlevel entry %s for shim %s : ",
+				       entry_id,
+				       source_shim);
+			return FALSE;
+		}
+		if (!fu_uefi_bootmgr_get_generation(current_entry, &current_generation, error)) {
+			g_prefix_error(error, "current sbatlevel entry %s: ", entry_id);
+			return FALSE;
+		}
```

Comment:
maybe we could have a `guint64 fu_csv_entry_get_value_by_column_id_uint64(FuCsvEntry *self, const gchar *column_id)` instead of this?